### PR TITLE
objclass: modify omap_get_{keys,vals} api

### PR DESCRIPTION
**Pull request by yehudasa**
add the truncated param (as with v2 of the corresponding librados api).

@liewegas this was only compile tested, need to go through rgw, rbd suites at least.
@dillaman note changes in cls_rbd and cls_journal. Looking at either it seems to me that we may want to think whether omap use there is scalable (unbounded listings).
@bassam see changes in cls_kvs
@noahdesu adjusted cls_lua, but you may want to incorporate new param to lua client

### CODE REVIEW

**Comment by dillaman** [comment]
@yehudasa where is the corresponding OSD change? I am confused by the param name of ```truncated``` -- I initially interpreted it as "your request didn't take up the full number of keys you asked for (i.e. it's a truncated result)" but you are actually using it as "there is more data to request". Can this parameter name be renamed so that it's clear that "more data exists"?

**Comment by yehudasa** [comment]
@dillaman `truncated` reflects the name of the param in the raw rados op, although librados uses `more` which might be less confusing. We were using `truncated` elsewhere before, but I'm not too attached to it. The objclass related changes are in https://github.com/ceph/ceph/pull/16352/commits/3cf7deb80deca89c49209ccec181a8afc4f9d8a7, there are no other osd changes (previous changes went in a while back).

**Comment by dillaman** [comment]
@yehudasa ah -- definitely a terrible name so it would be good to fix it

**Comment by yehudasa** [comment]
@dillaman renamed `truncated` to `more`, rebased

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-07-26_09:36:13-rados-wip-kefu-testing2-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2017-07-26_14:52:23-rados-wip-kefu-testing2-distro-basic-smithi/

**Comment by tchaikov** [comment]
@liewegas good to merge?

**Comment by liewegas** [comment]
retest this please

**Comment by tchaikov** [comment]
```
[----------] 2 tests from TestInstanceWatcher
[ RUN      ] TestInstanceWatcher.InitShutdown


The following tests FAILED:
	  1 - run-rbd-unit-tests.sh (Timeout)
	145 - unittest_rbd_mirror (Timeout)
Errors while running CTest
Build step 'Execute shell' marked build as failure
```

retest this please.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-07-27_04:52:14-rados-wip-kefu-testing-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2017-07-28_02:24:35-rados-wip-kefu-testing-distro-basic-mira/

**Comment by tchaikov** [comment]
```
[ RUN      ] TestInstanceWatcher.InitShutdown


98% tests passed, 3 tests failed out of 145

Total Test time (real) = 3748.41 sec
```

@yehudasa this test constantly times out in the jenkins' "make check" run, could you try to run it locally to see if it's reproducible in your testbed?

**Comment by liewegas** [comment]
@tchaikov, @yehudasa is out for 2 more weeks.  maybe @cbodley can look?

**Comment by cbodley** [comment]
i can reproduce this hang locally with:
> $ CEPH_LIB=lib bin/unittest_rbd_mirror --gtest_filter=TestInstanceWatcher.InitShutdown

`instances_list()` in cls_rbd.cc is going into an infinite loop because `cls_cxx_map_get_vals()` is returning with `vals.empty()` and `more == true`. will debug further

File: src/test/librados_test_stub/TestMemIoCtxImpl.cc [hunk 1]
```diff
@@ -209,9 +210,20 @@ int TestMemIoCtxImpl::omap_get_vals(const std::string& oid,
     }
     ++it;
   }
+  if (pmore) {
+    *pmore = (it != omap.end());
+  }
```
Comment (cbodley) [inline src/test/librados_test_stub/TestMemIoCtxImpl.cc]: missing `*pmore = false;` in early return above where `o_it == m_pool->file_omaps.end()`

**Comment by cbodley** [comment]
@liewegas `unittest_rbd_mirror` passes with the addition of https://github.com/ceph/ceph-ci/commit/5b14352ccd711765132cd99fb2fe57695d501f9c. i pushed a ceph-ci branch to https://shaman.ceph.com/builds/ceph/wip-objclass-omap/ - feel free to schedule the appropriate suites

**Comment by tchaikov** [comment]
@cbodley i will create another PR with your change so we can have a green "make check" run. `TestMemIoCtxImpl.cc` is not exercised by any test other than`unittest_journal` and `unittest_rbd_mirror`, so once the "make check" passes. this PR (actually the PR with your change) will be good to merge.

**Comment by tchaikov** [comment]
closing in favor of #16667

**Pull request closed without merging**